vpr/src/util/vpr_utils.cpp

```diff
@@ -2267,27 +2279,79 @@ int get_sub_tile_physical_pin(int sub_tile_index,
+                  logical_block->name, physical_tile->name);
+    }
+
+    return sub_tile_index;
+}
+
 int get_physical_pin(t_physical_tile_type_ptr physical_tile,
                      t_logical_block_type_ptr logical_block,
                      int pin) {
+    int sub_tile_index = get_logical_block_physical_sub_tile_index(physical_tile, logical_block);
+
+    if (sub_tile_index == OPEN) {
+        VPR_ERROR(VPR_ERROR_OTHER,
+                  "Couldn't find the corresponding physical tile type pin of the logical block type pin %d.",
+                  pin);
+    }
+
+    int sub_tile_physical_pin = get_sub_tile_physical_pin(sub_tile_index, physical_tile, logical_block, pin);
+    return physical_tile->sub_tiles[sub_tile_index].sub_tile_to_tile_pin_indices[sub_tile_physical_pin];
+}
+
+int get_logical_block_physical_sub_tile_index(t_physical_tile_type_ptr physical_tile,
+                                              t_logical_block_type_ptr logical_block,
+                                              int sub_tile_capacity) {
     int sub_tile_index = OPEN;
     for (const auto& sub_tile : physical_tile->sub_tiles) {
         auto eq_sites = sub_tile.equivalent_sites;
         auto it = std::find(eq_sites.begin(), eq_sites.end(), logical_block);
-        if (it != eq_sites.end()) {
+        if (it != eq_sites.end()
+            && (sub_tile.capacity.is_in_range(sub_tile_capacity))) {
             sub_tile_index = sub_tile.index;
+            break;
         }
     }
 
     if (sub_tile_index == OPEN) {
         VPR_ERROR(VPR_ERROR_OTHER,
-                  "Found no instances of logical block type '%s' within physical tile type '%s'. "
+                  "Found no instances of logical block type '%s' within physical tile type '%s'. ",
+                  logical_block->name, physical_tile->name);
+    }
+
+    return sub_tile_index;
+}
+
+int get_physical_pin(t_physical_tile_type_ptr physical_tile,
+                     t_logical_block_type_ptr logical_block,
+                     int sub_tile_capacity,
+                     int pin) {
+    int sub_tile_index = get_logical_block_physical_sub_tile_index(physical_tile, logical_block, sub_tile_capacity);
+
+    if (sub_tile_index == OPEN) {
+        VPR_ERROR(VPR_ERROR_OTHER,
                   "Couldn't find the corresponding physical tile type pin of the logical block type pin %d.",
-                  logical_block->name, physical_tile->name, pin);
+                  pin);
     }
 
     int sub_tile_physical_pin = get_sub_tile_physical_pin(sub_tile_index, physical_tile, logical_block, pin);
-    return physical_tile->sub_tiles[sub_tile_index].sub_tile_to_tile_pin_indices[sub_tile_physical_pin];
+    return (sub_tile_capacity - physical_tile->sub_tiles[sub_tile_index].capacity.low) * logical_block->pb_type->num_pins
+           + physical_tile->sub_tiles[sub_tile_index].sub_tile_to_tile_pin_indices[sub_tile_physical_pin];
```

Comment:
I think this calculation might lead to issues in case the logical block is a subset of the sub tile. In other words, if we use an equivalent logical block which has fewer number of pins, but can still fit in the sub tile, the returned pin would be invalid.

I think here there should be a similar calculation as in here: https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/37095c742471af2e2c8d90faf10ab13156a47942/vpr/src/util/vpr_utils.cpp#L2166-L2183

Basically the needed variables are:
- max_num_block_pins: which is the number of pins within one capacity instance of the sub_tile
- relative_capacity: which is indeed `sub_tile_capacity - physical_tile->sub_tiles[sub_tile_index].capacity.low`

Those variables can be used to retrieve the correct physical pin from the `sub_tile_to_tile_pin_indices` map.

Reply:
Looking at this PR code, I think that the modified `get_physical_pin` function gets called only once here: https://github.com/verilog-to-routing/vtr-verilog-to-routing/blob/95e704a9ba6b4590becadd5355b55cb8129ba5ef/vpr/src/pack/post_routing_pb_pin_fixup.cpp#L118

I see that there actually is a cluster block id at disposal. Basically, you could potentially use it IMO, as its purpose within the pin mapping utils function is to retrieve the location of that cluster block in the device.

Anyhow, I think that getting the max number of pins from the logical block might not be correct. Basically, you have all the information in the `sub_tile`, which you can use to get the maximum number of pins for each `sub_tile` location.

E.g.:
```
int max_num_block_pins = sub_tile.num_phy_pins / sub_tile.capacity.total();
```

The fact is that, you can always have the biggest logical block that fits in a sub tile that does not occupy all the physical pins that are available. If you get the `maximum number of pins` information from the logical block, you might end up in the same issue of mapping the logical pin at a wrong physical pin location.

Reply:
> I see that there actually is a cluster block id at disposal. Basically, you could potentially use it IMO, as its purpose within the pin mapping utils function is to retrieve the location of that cluster block in the device.

I need to correct this. Actually, using the clusterBlockId won't work as the placement has not been run yet when running the post routing cluster block fixup (correct?). If I am wrong and the block has been already placed, than you could use the already existing utility functions

Reply:
- Thanks for the advice on finding the proper number of pins for the subtile. I have adapted the code accordingly. 
- You are right regarding the use of `ClusterBlockId`. I can introduce it as an input argument but I have another consideration here. This function is also used in OpenFPGA when building the Verilog netlists modeling an FPGA fabric itself. As a fabric builder, there should be no dependency on mapping results. That is why I am trying to make the function only requiring device-level information. IMHO, the function is more general-purpose in that shape, and it can be used more widely. Feel free to let me know your vision here.